[Filestream] Expand take-over mode for Filestream inputs and support dynamic inputs (#42624)
elastic/beats: docs/reference/filebeat/filebeat-input-filestream.md

To remove the state of previously harvested files from the registry file, use the `clean_inactive` configuration option.

## Take over [filebeat-input-filestream-take-over]

> When `take_over` is enabled, this `filestream` input will take over
> states from the [`log`](/reference/filebeat/filebeat-input-log.md) input
> or other `filestream` inputs. Only states of files being actively
> harvested by this input are taken over.

Reviewer: I don't it would be entirely clear to the users what states are. It's more an implementation detail.

```suggestion
When `take_over` is enabled, this `filestream` input will take over ingestion of all
files from [`log`](/reference/filebeat/filebeat-input-log.md) or [`filestream`](/reference/filebeat/filebeat-input-filestream.md) inputs that are matched by the `paths` of this `filestream` input.
```

Author: I believe they have some idea, but I agree we can word it better.

Your version seems to be missing the fact that only the states of existing files are migrated, I'll re-word it later today.

> To take over states from the log input, simply set `take_over.enabled:
> true`. To take over states from other `filestream` inputs, set
> `take_over.enabled: true` and set `take_over.from_ids` as list of
> previous `filestream` IDs you want to migrate states from.

Reviewer: ```suggestion
To take over files from a log input, simply set `take_over.enabled:
true` and add a matching path(s) to the `paths` list.

To take over states from other `filestream` inputs, set `take_over.enabled: true`, add matching paths to the `paths` list and set `take_over.from_ids` to a list of existing `filestream` IDs you want to migrate files from.

When `take_over.from_ids` is set, files are not taken over from `log` inputs. The migration is limited to `filestream` inputs only.
```

Author: Overall I like your suggestion, I just wonder if we really need to mention the paths. That's the most basic configuration of Filestream, paths has got to be set to the files the user wants ingested. It feels really redundant for me.

I like all other changes.

Reviewer: I'm afraid that if we don't mention paths users would think it's enough to just set the `take_over.from_ids` and everything would magically work. From my standpoint it's better to be explicit about the complete requirements.

Author: I still believe it is redundant to mention that `paths` needs to be set. Specially as a step for the migration. I'll re-word it to mention the files we want to take over need to match the globs on `paths`.

```yaml
take_over:
  enabled: true
  from_ids: ["foo", "bar"] # omit to take over from the log input
```
:::{important}

> Take over can only be used if the inputs it is taking states from are
> not active any more. If old inputs harvesting the file are still
> active, the data ingestion will be inconsistent (data duplication or
> data loss will occur).

Reviewer: ```suggestion
The `take over` mode can work correctly only if the source (taken from) inputs are no longer active. If source inputs are still harvesting the files which are being migrated, it will lead to data duplication and in some cases might cause data loss.
```

:::

::::{important}
`take_over.enabled: true` requires the `filestream` to have a unique ID.
::::

This `take over` mode was created to enable smooth migration from
deprecated `log` inputs to the new `filestream` inputs and to allow
changing `filestream` input IDs without data re-ingestion.

See [*Migrate `log` input configurations to `filestream`*](/reference/filebeat/migrate-to-filestream.md) for more details about the migration process.

::::{warning}
The `take over` mode is still in beta, however, it should be generally safe to use.
::::

### Limitations
Take over can only migrate states from existing files that are not
ignored during the `filestream` input start up. Once the input is
ingesting data, if a new file appears, `filestream` will not try to

> migrate its state.

Reviewer: I think it's not accurate.

In order to migrate the files from one input to another (regardless the input type) we need to put matching path/glob on the `paths` list. This means if a new file appears, the new input should pick it up just fine. Did I get it wrong?

Author: You're correct that if a new file matching the configured globs appears, Filestream will pick it up and start harvesting it.

However, the states take over, actually all store manipulation (store clean up, migrating identifiers from `.global` to the actual ID and file identity migration) happen during the prospector initialisation. The first thing the `Init` method does is to scan the file system and get a [list of files](https://github.com/belimawr/beats/blob/77654140e4e0fd7a0d3ca9af96a980d13126648e/filebeat/input/filestream/prospector.go#L85-L90), then when going through the registry entries, the paths are compared with the list of files found in the filesystem.

This is important so we can migrate the file identity. The Log input does not support fingerprint and fingerprint now is the default for Filestream, so we need access to the actual file during the migraiton.

I've been thinking a lot about this limitation, however I cannot see it being an actual issue for a normal operation of Filebeat and log rotation strategies. A file that has been fully (or partially) ingested and has been removed (hence it would not be found by the initial scan `Init` does) would not be restored (or somehow appear again).

Doing this during the prospector initialisation ensures the rest of the code base will have a consistent to the registry/file offset, by the time the file is open for reading, the entries in the store have already been migrated.

Reviewer: My point was rather about this part:

> Once the input is ingesting data, if a new file appears, `filestream` will not try to migrate its state.

The statement is correct but the migration is not needed in this case. 

Because the `paths` in the new input must match the new file (in order to trigger the migration in the first place), it's going to be picked up as a new file and ingested by the new input just fine without any state migration. The end result for the customer is the same – the new file is ingested. That's why I think this section is redundant, it may only create confusion and a feeling of false limitation.

Given the old input is not running at the time of migration (which we warn not to do), between the time when the "take over" migration starts and the prospector starts watching files, no state changes would happen.

Author: I see, thanks!

I'll remove the last bit
> Once the input is ingesting data, if a new file appears, `filestream` will not try to migrate its state.

```yaml
take_over:
  enabled: true
  from_ids: ["foo", "bar"] # omit to take over from the log input
```
:::{important}
The `take over` mode can work correctly only if the source (taken from) inputs are no longer active. If source inputs are still harvesting the files which are being migrated, it will lead to data duplication and in some cases might cause data loss.
:::

::::{important}
`take_over.enabled: true` requires the `filestream` to have a unique ID.
::::

This `take over` mode was created to enable smooth migration from
deprecated `log` inputs to the new `filestream` inputs and to allow
changing `filestream` input IDs without data re-ingestion.

See [*Migrate `log` input configurations to `filestream`*](/reference/filebeat/migrate-to-filestream.md) for more details about the migration process.

::::{warning}
The `take over` mode is still in beta, however, it should be generally safe to use.
::::

### Limitations
Take over can only migrate states from existing files that are not
ignored during the `filestream` input start up. Once the input is
ingesting data, if a new file appears, `filestream` will not try to

> migrate its state.

Author: > It's for the case the old input is left running, detects a new file and creates a state for the new file, right?

I'm not sure I quite get what you meant there. No, this has nothing to do with the old input. The code for take over is in the prospector `Init` method, which only runs when the input is starting up. This is a safe place because the state has not been read/propagated to the bits of code that will keep copies of the offset.

> In that case filestream will also detect the new file and ingest it normally. Is it?

Yes, any file that appears after the prospector `Init` runs, will be detected and handled normally.
